README.md
# emacs-mail-client

This repo contains my hacks to read email in GNU Emacs.

The first question is: why?  The short answer: I'm much more
productive with this system than anything I've seen other people use.
That doesn't mean there isn't something out there more powerful, but I
just haven't seen it.

The long answer:

Back in the 90's I was using GNU Emacs + MH-E and I needed a way to
manage the volume of email I was getting.  Chris Richardson and I
cooked up a filtering system based on Perl.  It worked by sitting in
front of **inc** (the MH way of reading new mail) that would sort new
mail into different *inboxes*.  The rules were written in Perl and
were fairly concise.

I used the Perl-based filter for some time.  Then later in the 90's
Ahmon Dancy wrote a Common Lisp version of the Perl script, and
**mailfilter** was born.  The Common Lisp version features rules in
Common Lisp, obviously.  Look at *mailfilter.cl* in the repo to see
what the rules look like.  This is the actual rules I use for my
personal email.  The ones for work are a lot more complex, but contain
things which I don't want to make public.

I used the system in this state for another 15 years, when I really
needed something better to manage the large number of balls I was
keeping in the air.  The problem: my **+inbox** tended to grow large
over short periods of time and I'd lose track of all the messages in
my various other inboxes.  Not only that, I would have to periodically
spend several hours cleaning up my **+inbox**.  It was really painful.

The first thing I did was write an Emacs mode for looking at the
inboxes.  Instead of looking at my **+inbox**, I could look at a
summary of my inboxes:

    +inbox            1 new                 170 old  
    +inbox-gmail                34 unread    35 old  
    +inbox-spam       1 new    233 unread   237 old  
    +inbox-junk       1 new    151 unread   151 old  

    ; *********** TODO inboxes ********************
    +inbox-q                                 35 old  
    +inbox-fin                               41 old  
    +inbox-todo                              37 old  
    +inbox-sa                                47 old  
    +inbox-pbx                               11 old  
    +inbox-someday                           77 old  
    +inbox-pend                               1 old  

    ; *********** Hobby inboxes********************
    +inbox-music                             26 old  
    +inbox-photo                             26 old  
    +inbox-fun                               21 old  
    +inbox-travel                             1 old  

    ; *********** Mailing list inboxes ************
    +inbox-emacs      1 new    298 unread   307 old  
    +inbox-git                 178 unread   178 old  
    +inbox-blink                10 unread    10 old  
    +inbox-cygwin              104 unread   131 old  
    +inbox-mh                                 2 old  
    +inbox-nmh                                4 old  
    +inbox-ps                    1 unread     1 old  

A single key would *go into* a folder and I could process those new,
unread or old emails.

Notice I said *could* above.  The problem was, I didn't make it the
default mode.  It was a big step moving to making the *inboxes mode*
the default, but it was an important one.

Once I started using the *inboxes mode*, I ran into another problem:
it was hard to move messages to various inboxes because new email on
those conversations kept coming back to my **+inbox**.  So, I wrote
support to *move conversations* to a folder so that new emails would
automatically show up there.

Lastly, some time before this I had written an MH-E hack that would
allow me to highlight messages.  I could mark them as *important* and
they would be highlighted in yellow in whatever folder they were in.
I used nmh sequences for this, but, unfortunately, when a message is
refiled from one folder to another, the sequence information is lost.
The *nmh-workers* mailing list was helpful in finding a solution:
a *refile hook* to make sure the newly moved message was in the same
sequence.

## Dependencies

First, this doesn't work on Windows.  I've tested it on many versions
of Fedora up to version 17.  It will very likely work on CentOS, RHEL,
Ubuntu, and many other flavors of UNIX.  The software originally ran
on Solaris.

This repo depends on these parts:

 * [nmh](http://www.nongnu.org/nmh/) 1.5 :: this is usually provided by the operating system
 * [mairix](http://www.rpcurnow.force9.co.uk/mairix/) :: Mairix is used for indexing and searching messages
 * [MH-E](http://mh-e.sourceforge.net/) 8.5 :: the Emacs interface to MH/nmh
 * [mailfilter](https://github.com/franzinc/mailfilter) 2.x :: the mail filter that sits on top of MH/nmh
 * [Allegro Common Lisp](http://www.franz.com) :: because *mailfilter*
   is written in Common Lisp, you need a Common Lisp.  Another Common
   Lisp might do, but I've never tested anything but ACL.  I will be
   providing binaries for *mailfilter* at some point in the future.

The rest is included in this repo.

## Installation

### MH/nmh

I won't talk about nmh installation.  Get that installed and come back
here.

### MH-E

There is included in this repo a copy of MH-E 8.5.  It is unchanged
from the official release and is included here for convenience.

To compile the `.el` files, open an shell and execute these commands:

    $ cd .../emacs-mail-client
    $ make

This will use the GNU Emacs binary installed on your system to compile
the `.el` files to `.elc` files for faster operation.

### mailfilter

[Installation instructions.](https://github.com/franzinc/mailfilter#readme)

### $HOME/.mh_profile

You need these additions to your `~/.mh_profile`:

    #: Immediately remove items from some folders, moving the rest
    #: to the +trash folder -- customize to your needs.  See also
    #: the mh-cleanup.sh script below.
    rmmproc: /home/layer/src/emacs/emacs-mail-client/mh-rmmproc.sh
    send: -msgid
    #: so messages in the `important' sequence stay in it after being moved
    ref-hook: /home/layer/src/emacs/emacs-mail-client/mh-ref-hook.sh

Of course, use the actual path to this repo on your system.

### $HOME/.mailfilter.cl

Create your own from what's in this repo.  [You can see what I have in
mine.](https://raw.github.com/e40/emacs-mail-client/master/mailfilter.cl)
Start out simple and grow it.

### GNU Emacs

First, you need to load `dkl-boot.el`.  This contains a bunch of
functions which are later used.  This is how I load it from my
`~/.emacs`:

    (require 'bytecomp)
    ;; Must load this first because it contains various support functions
    (let* ((el (format "%semacs-mail-client/dkl-boot.el"
                       (file-name-directory load-file-name)))
           (elc (byte-compile-dest-file el)))
      (cond ((file-newer-than-file-p elc el) (load elc))
            (t (byte-compile-file el t))))

Then, load `dkl-mh-e.el` from this repo.  I do it this way:

    (load (format "%s/emacs-mail-client/dkl-mh-e.el" *my-elib*))

`*my-elib*` is defined to point to `~/src/emacs/`.  You need to define
keybindings for the entry point functions, and this is what I do:

    ;; visit a specific inbox:
    (define-key ctl-x-map "a" 'dkl:mh-scan-inboxes)

    ;; compose an email:
    (define-key ctl-x-map "c" 'mh-smail)

    ;; read email:
    (define-key ctl-x-map "i" 'dkl:mh-rmail)

    ;; visit a specific MH/nmh folder:
    ;;  (I call it with an argument so it asks which MH folder to visit)
    (define-key ctl-x-map "s" (lambda ()
                                (interactive)
                                (mh-rmail 1))))

    ;; Move conversations with a single binding.  Like
    ;; mh-thread-refile.
    (define-key mh-folder-mode-map "\e^" 'dkl:mh-move-conversation)

    ;; Mark messages importand/unimportant:
    (define-key mh-folder-mode-map "*" 'my-mark-important)
    (define-key mh-folder-mode-map "_" 'my-mark-unimportant)

The above are the basic key bindings.  

In summary:

| Keybinding | Action                   | Comments                       |
|:----------:|:-------------------------|:-------------------------------|
|`C-x a`     | visit one of the inboxes |folder read from minibuffer     |
|`C-x c`     | compose a message        |creates a new draft             |
|`C-x i`     | show `*Inboxes*` buffer  |w/prefix arg, inc to +inbox     |
|`C-x s`     | visit a folder           |folder read from minibuffer     |
|`ESC ^`     | move conversation        |folder read from minibuffer     |
|`*`         | mark message important   |                                |
|`_`         | mark message unimportant |                                |

*Inboxes* are defined to be folders named with the prefix `inbox-`.
You can create an inbox by just refiling a message to it, and creating
it in the process, or having `incfilter` create the message there.
(Will `incfilter` create a folder or does it need to already exist?)

### Indexing email with Mairix

configuring
cron job

    # Mairix is fast, so index once an hour
    0 * * * * cd mail; ./index.sh

### cleaning up

    # cleanup mail at 4:01am
    1 4 * * * cd mail; ./cleanup


mh-e-8.4/ChangeLog
2012-11-25  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.4.

	* README: Update for release 8.4.

2012-11-24  Bill Wohler  <wohler@newt.com>

	* test-mh-e: Use EMACS_SRC environment variable to override default
	emacs/trunk branch (with emacs/mh-e, for example). Use it within
	the rest of the script and pass it to make.

	* mh-unit.el: Update copyright.
	(mh-unit): Use EMACS_SRC environment variable to override default
	emacs/trunk branch (with emacs/mh-e, for example).
	(mh-unit-test-x-image-url-cache-canonicalize): Update expected
	strings for Emacs 24. It appears that url-hexify-string now
	converts ! to %21 and uses capital letters for hexadecimal digits.

2011-11-20  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.3.1.

	* README: Update for release 8.3.1.

2011-09-19  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.3.

	* README: Update for release 8.3.

2011-07-17  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.2.93.

	* README: Update for release 8.2.93.

	* test-mh-e: Add script used for testing MH-E.

2011-07-16  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.2.92.

	* README: Update for release 8.2.92.

	* mh-xemacs.el (mh-modeline-glyph): Update glyph so that it does
	not cause errors on terminal window even when created on a window
	system.  Thanks to Henrique Martins and Mike Kupfer.

2011-07-13  Bill Wohler  <wohler@newt.com>

	* mh-xemacs.el (mh-modeline-glyph): Revert previous change since
	it didn't help on XEmacs.

2011-07-11  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.2.91.

	* README: Update for release 8.2.91.

	* mh-xemacs.el (mh-modeline-glyph): Don't define in terminal
	windows to avoid errors. Thanks to Henrique Martins.

	* README: Add trunk to `load-path.' Thanks to Henrique Martins.

2011-07-10  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.2.90.

	* README: Update for release 8.2.90.

	* COPYING, Makefile, mh-unit.el, mh-xemacs.el, release-utils:
	Update to GNU GPL version 3, use date ranges in copyright.

	* release-utils (tag): Remove errant @ character. Call bzr unbind
	before bzr bind, since bzr bind fails if branch already bound.

2011-07-09  Bill Wohler  <wohler@newt.com>

	* mh-unit.el: Use eval-and-compile instead of eval-when-compile
	since mh-require-cl is needed at evaluation time too.
	(mh-unit): Append trunk to emacs paths.
	(mh-unit-construct-call-graph)
	(mh-unit-test-normalize-folder-name)
	(mh-unit-test-sub-folders-actual)
	(mh-unit-test-sub-folders): Use OK at end of message instead of
	done.
	(mh-unit-test-folder-completion-function): Update tests per
	Stefan's recent changes to MH-E's folder completion.
	(mh-unit-equal): Display expected string first to be consistent
	with JUnit, for example.

2011-07-04  Bill Wohler  <wohler@newt.com>

	* release-utils (tag): Explicitly bind to branch so that tag
	affects upstream branch.

2011-07-03  Bill Wohler  <wohler@newt.com>

	* Makefile: Put all copyrights written to files in ranges and add
	2011 to range.

2011-07-02  Bill Wohler  <wohler@newt.com>

	* release-utils (distCheck): Remove debugging code. Use -ne in
	comparison of numbers.

2011-06-26  Bill Wohler  <wohler@newt.com>

	* README: Replace references to CVS with Bazaar.

	* release-utils (dist, tag, variable_changes): Update from cvs to
	Bazaar.

	* Makefile (CVSROOT, CVSLOGIN, CVSMETHOD) Replace with URL and
	DEV_URL.
	(MH_E_SRC, MH_E_SRC_NONCOMPILED, MH_E_LOADDEFS)
	(MH_E_AUTOLOADS, MH_E_CUSTOM_LOAD, MH_E_IMG, MH_E_ETC)
	(MH_E_MAINT): Define and use EMACS_SRC for emacs/trunk.
	(emacs, emacs-devel): Use bzr instead of cvs.
	(tag, dist): Update args for release-utils --tag.
	(backup): Remove. Obsolete.

2009-06-13  Bill Wohler	 <wohler@newt.com>

	Release MH-E version 8.2.

	* README: Update for release 8.2.

2008-08-18  Bill Wohler	 <wohler@newt.com>

	Release MH-E version 8.1.

	* README: Update for release 8.1.

2006-11-13  Bill Wohler  <wohler@newt.com>

	* mh-unit.el (mh-unit-test-x-image-url-cache-canonicalize): Add
	test for invalid Window characters (closes SF #1396499).

2006-11-12  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.0.3.

	* README: Update for release 8.0.3.

	* mh-unit.el (mh-unit-test-x-image-url-cache-canonicalize): Hex
	equivalents now use lowercase.

2006-07-03  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.0.2.

	* README: Update for release 8.0.2.

2006-07-03  Bill Wohler  <wohler@newt.com>

	* mh-xemacs.el: (mh-icon-image): New function. Returns XPM from
	icon file.
	(mh-xemacs-icon-list): New constant. List of icons used by MH-E.
	(mh-xemacs-icon-map): Initialize in place, using mh-icon-image to
	read images from icon files.
	(mh-modeline-glyph): Use mh-logo file instead of mh-modeline-logo
	constant.
	(mh-modeline-logo, mh-xemacs-attach-icon, mh-xemacs-close-icon)
	(mh-xemacs-copy-icon, mh-xemacs-cut-icon)
	(mh-xemacs-execute-icon, mh-xemacs-fld-open-icon)
	(mh-xemacs-help-icon, mh-xemacs-highlight-icon)
	(mh-xemacs-left-arrow-icon, mh-xemacs-mail-icon)
	(mh-xemacs-mail-alias-icon, mh-xemacs-mail-compose-icon)
	(mh-xemacs-mail-refile-icon, mh-xemacs-mail-repack-icon)
	(mh-xemacs-mail-reply-icon, mh-xemacs-mail-reply-all-icon)
	(mh-xemacs-reply-from-icon, mh-xemacs-reply-to-icon)
	(mh-xemacs-mail-send-icon, mh-xemacs-page-down-icon)
	(mh-xemacs-paste-icon, mh-xemacs-preferences-icon)
	(mh-xemacs-refresh-icon, mh-xemacs-right-arrow-icon)
	(mh-xemacs-save-icon, mh-xemacs-search-icon)
	(mh-xemacs-show-icon, mh-xemacs-spell-icon, mh-xemacs-undo-icon)
	(mh-xemacs-widen-icon): Delete.

	* Makefile (MH_E_IMG): Add save.pbm and save.xpm.

2006-07-03  Ted Phelps  <phelps@gnusto.com>  (tiny change)

	(mh-xemacs-unknown-icon): New icon. Used if icons can't be found.

2006-06-20  Bill Wohler  <wohler@newt.com>

	Release MH-E version 8.0.1.

	* README: Update for release 8.0.1.

2006-06-20  Bill Wohler  <wohler@newt.com>

	* README: Depend on GNU mailutils 1.0 (not yet released) and
	higher due to a bug in folder.


dkl-mh-e.el

;; Add one of these to ~/.emacs
;;;(custom-set-variables '(mh-identity-default "Home"))
;;;(custom-set-variables '(mh-identity-default "Work"))

;;;(setq load-path
;;;  (cons (expand-file-name "~layer/src/emacs/gnus.git/lisp/") load-path))
;;;(require 'gnus-load)

(push
 (format "%smh-e-8.5/emacs/trunk/lisp/mh-e/"
	 (file-name-directory load-file-name))
 load-path)

(dkl:safe-load "message")
(dkl:safe-load "mh-e")

(when (string= "8.4" mh-version)
  (error "MH-E 8.4 is buggy.  Don't use."))

(dkl:safe-require 'mh-comp)
(dkl:safe-require 'mh-mime)
(dkl:safe-require 'mh-utils)
(dkl:safe-require 'mh-alias)
(dkl:safe-require 'mh-junk)
;; MH-E 8.0 needs this (for mh-to-fcc):
(dkl:safe-require 'mh-letter)
(dkl:safe-require 'mh-folder)
(dkl:safe-require 'mh-utils) ;; make sure mh-folder-completion-map is defined
(dkl:safe-require 'mh-thread)

(dkl:byte-compile-file (format "%sdkl-mh-e-fixes.el"
			       (file-name-directory load-file-name))
		       t)
(dkl:byte-compile-file (format "%sdkl-mh-mailfilter.el"
			       (file-name-directory load-file-name))
		       t)

;;;(when (file-exists-p "/usr/bin/mh")
;;;  (custom-set-variables
;;;   '(mh-progs "/usr/bin/mh/")
;;;   '(mh-lib "/usr/lib/mh/")
;;;   '(mh-lib-progs mh-lib)))

(custom-set-variables
 ;; In search of a good renderer for HTML emails!  Oh my!
 '(mm-text-html-renderer
   ;;'w3m ;; emacs-w3m -- not installed
   ;;'w3 ;; Emacs/W3 -- not installed
   ;; 'w3m-standalone ;; w3m -- ugly, no pics, no links, fast
   ;;'links ;; ugliest, no pics, fast, links at bottom
   ;;'lynx ;; OK, no pics, no links, fast
   ;;'html2text -- unusable
   ;;nil ;; external viewer (default web browser)
   
   'shr ;; gnus, broken in 24.[12] w/o patch, sometimes hang w/patch
   )

 '(mh-identity-list
   ;; ~/.emacs sets mh-identity-default to one of these:
   (quote (("Home" (("From" . "Kevin Layer <layer@known.net>")
		    ("Fcc" . "+outbox")))
	   ("Work" (("From" . "Kevin Layer <layer@franz.com>")
		    ;;("Organization" . "Franz, Inc.")
		    ("Fcc" . "+outbox"))))))
 '(mm-attachment-file-modes 420)	; mode 644 not 600!
 (list 'mh-inc-prog
       (cond ((file-exists-p "/usr/local/incfilter") "/usr/local/incfilter")
	     ((file-exists-p "/usr/fi/incfilter") "/usr/fi/incfilter")
	     (t "incfilter")))
 (list 'mh-alias-local-users
       (if (at-work-on-unix-p)
	   "ypcat passwd"
	 ;; use /etc/passwd:
	 t))
 '(mh-invisible-header-fields-default
   ;; mh has a built-in list (mh-invisible-header-fields-internal) of
   ;; headers that it will not show.  Entries in
   ;; mh-invisible-header-fields-default will be removed from that list.
   ;; Make sure you specify removals in the same way that they are listed
   ;; in mh-invisible-header-fields-internal.  I.e., punctuation (colons)
   ;; and case matter.
   '("Message-Id:")) 
 '(mh-invisible-header-fields ;; hide 
   '("DKIM-Signature" "Comment" "DomainKey-" "X-" "In-" "Organization"
     "References" "Reply-To" "Sender" "User-" "Mail-" "Bh-Id:" "Class:"))
 '(mh-refile-preserves-sequences-flag
   ;; 9/6/04:
   ;; Done after seeing discussion on the mh-e-users mailing list of the
   ;; error message:
   ;;   refile: message 55 doesn't exist
   nil)
 '(mh-compose-forward-as-mime-flag nil) ;; don't forward as mime attachment
 '(mh-mime-save-parts-directory "~/incoming/")
 '(mh-search-program 'mairix)
 '(mh-show-use-xface-flag nil) ;; t is too slow...
 '(mh-junk-disposition
   ;; Stuff marked as `junk' will be moved to this folder
   "+inbox-spam")
 '(mh-junk-program 'spamassassin)
 '(mh-nmh-flag t)			;for Linux systems, very important
 '(mh-show-threads-flag t)
 '(mh-large-folder 500)
 '(mh-recenter-summary-flag t)
 '(mh-signature-file-name "~/.sig")
 '(mh-show-buffer-mode-line-buffer-id "%s/%d")
 '(mh-redist-full-contents
   ;; doesn't work if `t' if you've replied to the message already
   nil)
 '(mh-clean-message-header-flag t)
 '(mh-lpr-command-format "nenscript -Plw2 -h")
 '(mh-ins-buf-prefix ">> ")
 '(mh-delete-yanked-msg-window t)
 '(mh-reply-default-reply-to "cc"))


(defun my-mh-compose-letter-function (to subject cc)
  ;; automatically fix bh header when report id in subject
  (when (and (fboundp 'bh=field-value)
	     (or subject
		 ;; When we do mh-edit-again, subject is nil, but it exists
		 ;; in the already composed email, so grab it:
		 (setq subject (bh=field-value "^Subject: ")))
	     (boundp 'bh-report-regexp)
	     bh-report-regexp
	     (string-match bh-report-regexp subject))
    (bh-fix-mail-header
     (downcase (substring subject (match-beginning 1)
			  (match-end 1)))))
  
  (save-excursion
    (goto-char (point-min))
    (replace-string (char-to-string 160) " ")))

(add-hook 'mh-compose-letter-function 'my-mh-compose-letter-function)

(defun my-mh-show-mode-hook ()
  (save-excursion
    (let ((was-read-only buffer-read-only))
      (when buffer-read-only (toggle-read-only -1))
      (goto-char (point-min))
      (while (search-forward (char-to-string 160) nil t)
	(replace-match " " nil t))
      (when was-read-only (toggle-read-only 1)))))

(add-hook 'mh-show-mode-hook 'my-mh-show-mode-hook)

;;;(defun my-mh-compose-letter-test-hook (to subject cc)
;;;  (message "foo: %s" mh-sent-from-folder)
;;;  (when (string-match "^\\+inbox$" mh-sent-from-folder)
;;;    (mh-to-fcc "+foobar")))
;;;(add-hook 'mh-compose-letter-function 'my-mh-compose-letter-test-hook)

;;;(defun my-move-to-saved ()
;;;  (interactive)
;;;  (mh-refile-a-msg nil '+saved)
;;;  (mh-next-msg t))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Begin highlighting code

(defun my-mh-highlight-folder-hook-function ()
  (my-highlight-messages 'my-mh-highlight-folder-hook-function
			 mh-current-folder))

(add-hook 'mh-inc-folder-hook 'my-mh-highlight-folder-hook-function)

(defadvice mh-visit-folder (after my-highlight-messages
				  (folder &optional range index-data)
				  activate)
  (my-highlight-messages 'mh-visit-folder mh-current-folder (ad-get-arg 1)))

(defadvice mh-pack-folder (after my-highlight-messages (range) activate)
  (my-highlight-messages 'mh-pack-folder mh-current-folder (ad-get-arg 0)))

(defadvice mh-rescan-folder (after my-highlight-messages
				   (&optional range dont-exec-pending)
				   activate)
  (my-highlight-messages 'mh-rescan-folder mh-current-folder (ad-get-arg 1)))

(defadvice mh-execute-commands (after my-highlight-messages () activate)
  (my-highlight-messages 'mh-execute-commands mh-current-folder))

(defadvice mh-widen (after my-highlight-messages () activate)
  (my-highlight-messages 'mh-widen mh-current-folder))

;;;;;;;;;

(defface my-scan-line-highlight '((t
			      ;;; one of these:
				   (:background "yellow")
				   ;;(:foreground "DarkGreen")
				   ))
  "The face I use to highlight scan lines in MH-E."
  :group 'my-faces)


(defun my-highlight-messages (from folder &optional range)
;;;;used to debug redundant calls to this function
  ;;(message "my-highlight-messages from %s" from)
  (my-highlight-by-header mh-current-folder range)
  (my-highlight-by-sequence mh-current-folder range))


(defun my-highlight-by-header (folder range)
  (condition-case ()
      (my-highlight-by-header-1
       folder range "highlight" 'my-scan-line-highlight)
    (error nil)))

(defun my-highlight-by-header-1 (folder range header face)
;;;;used to debug redundant calls to this function
  ;;(message "my-highlight-by-header-1: folder=%s range=%s" folder range)
  (let ((output
	 (if range
	     (my-mh-exec-cmd "pick" folder range "--X-Layer-Filter" header)
	   (my-mh-exec-cmd "pick" folder "--X-Layer-Filter" header)))
	messages over)
    (if (string-match "pick: no messages match specification" output)
	nil
      ;; `messages' has the message numbers which we need to highlight.
      (save-excursion
	(setq messages (dkl:read-objects-from-string output))
	(dolist (msgn messages)
	  (mh-goto-msg msgn nil t)
	  (setq over (make-overlay
		      (progn (beginning-of-line) (point))
		      (progn (end-of-line) (point))
		      nil nil nil))
	  (overlay-put over 'face face))))))

;;;;;; can't remember why I needed the following
(defun my-mh-exec-cmd (&rest command)
  (save-excursion
    (let ((temp " *my-mh-temp*"))
      (set-buffer (get-buffer-create temp))
      (erase-buffer)
      (apply 'mh-exec-cmd-output (car command) nil (cdr command))
      (buffer-string))))


(defconst my-mh-important-msg-seq 'important)


(defun my-highlight-by-sequence (folder range)
  (condition-case ()
      (my-highlight-by-sequence-1
       folder range my-mh-important-msg-seq 'my-scan-line-highlight)
    (error nil)))

(defun my-highlight-by-sequence-1 (folder range sequence face)
  (let (messages over)
    (save-excursion
      (setq messages (mh-seq-to-msgs sequence))
      (dolist (msgn messages)
	(mh-goto-msg msgn nil t)
	(setq over (make-overlay
		    (progn (beginning-of-line) (point))
		    (progn (end-of-line) (point))
		    nil nil nil))
	(overlay-put over 'face face)))))



(defun my-unhighlight-message (msgn)
  (save-excursion
    (mh-goto-msg msgn nil t)
    (remove-overlays (progn (beginning-of-line) (point))
		     (progn (end-of-line) (point)))))


(defun my-mark-important (&optional msg folder)
  (interactive (list (mh-get-msg-num t) mh-current-folder))
  (when (not (member my-mh-important-msg-seq
		     (mh-seq-containing-msg msg nil)))
    (let ((msgs (list msg)))
      (mh-add-msgs-to-seq msgs my-mh-important-msg-seq)
      (my-highlight-by-sequence folder msgs)
      (mh-next-msg nil)
      (mh-define-sequence 'cur (list (mh-get-msg-num t))))))

(defun my-mark-unimportant ()
  (interactive)
  (let ((current-message (mh-get-msg-num t)))
    (when (member my-mh-important-msg-seq
		  (mh-seq-containing-msg current-message nil))
      (mh-delete-a-msg-from-seq current-message my-mh-important-msg-seq nil)
      (my-unhighlight-message current-message)
      (mh-next-msg nil)
      (mh-define-sequence 'cur (list (mh-get-msg-num t))))))

;;;; not needed yet
;;;(defun my-message-in-sequence-p (message seq)
;;;  (memql message (mh-seq-to-msgs seq)))

(defun my-move-to-end (move-thread)
  (interactive "P")
  (cond (move-thread
	 (let ((region
		;; list of (start-position end-position)
		(mh-thread-find-children))
	       (messages '()))
	   (when (not (and (consp region)
			   (eql 2 (length region))))
	     (error "region is not a list of length 2: %s" region))
	   (save-excursion
	     (my-map-lines-region
	      (first region)
	      (second region)
	      (lambda () (push (mh-get-msg-num t) messages)))
	     (setq messages (nreverse messages))
	     (dolist (message messages)
	       (my-move-to-end-current-message nil message)))
	   (mh-rescan-folder)))
	(t (my-move-to-end-current-message t))))

(defun my-move-to-end-current-message (rescan &optional message)
  (let* ((message (or message (mh-get-msg-num t)))
	 (sequences (mh-seq-containing-msg message
					   ;; don't include `cur' and
					   ;; internal nmh/MH-E sequences
					   nil)))
    ;; move to end
    ;;   refile 257 -nolink -src +inbox +inbox
    (mh-refile-msg message
		   (intern mh-current-folder)
		   t ;; dont-update-last-destination-flag
		   )
    
    (mh-execute-commands)
    
    ;; add message to sequences
    (dolist (sequence sequences)
      (mh-add-msgs-to-seq (mh-translate-range mh-current-folder "last")
			  sequence nil t))
    
    (when rescan (mh-rescan-folder))))

;; ...end highlighting code
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(make-variable-buffer-local 'smooth-scroll-margin)

(defun my-mh-folder-mode-hook-function ()
  (interactive)
  (setq smooth-scroll-margin 0) ;; >0 doesn't work well with folder mode
  (cd "~/")
  (font-lock-mode))
(add-hook 'mh-folder-mode-hook 'my-mh-folder-mode-hook-function)

(defun my-mh-letter-mode-hook-function0 ()
  (run-hooks 'text-mode-hook)
  (setq comment-start ">> "))

(defun my-mh-letter-mode-hook-function2 ()
  ;; check if To: has "lifemasters.com" in it, then
  ;; 1. Cc: Powell and Fred
  ;; 2. Fcc: +inbox
  (save-excursion
    (let ((case-fold-search t))
      (when (string-match "lifemasters\\.com" (mh-get-header-field "To"))
	(let ((auto-cc '("swright@lifemasters.com"
			 "phamner@lifemasters.com"
			 "cnorvell"
			 "dm"))
	      (cc (mh-get-header-field "Cc")))
	  (dolist (acc auto-cc)
	    (when (not (string-match acc cc))
	      (my-mh-insert-fields "Cc:" acc)))
	  (my-mh-insert-fields "Fcc:" "inbox"))))))

(defun my-mh-insert-fields (&rest name-values)
  "Insert the NAME-VALUES pairs in the current buffer.
If the field exists, append the value to it.
Do not insert any pairs whose value is the empty string."
  (let ((case-fold-search t))
    (while name-values
      (let ((field-name (car name-values))
            (value (car (cdr name-values))))
        (if (not (string-match "^.*:$" field-name))
            (setq field-name (concat field-name ":")))
        (cond ((equal value "")
               nil)
              ((mh-position-on-field field-name)
;;;;DKL: changed " " to ", "
               (insert ", " (or value "")))
              (t
               (insert field-name " " value "\n")))
        (setq name-values (cdr (cdr name-values)))))))

(add-hook 'mh-letter-mode-hook 'my-mh-letter-mode-hook-function0)
(add-hook 'mh-letter-mode-hook 'my-mh-letter-mode-hook-function2)

(condition-case ()
    (mh-find-path)
  (error nil))
